protocol/standard-l2-genesis.md

```diff
@@ -0,0 +1,185 @@
+given our development practices, but every time that we do want to do a release it results in a lot of time
+being spent trying to make sure that we are upgrading to a compatible set of contracts.
+
+# Proposed Solution
+
+<!-- A high level overview of the proposed solution.
+When there are multiple alternatives there should be an explanation
+of why one solution was picked over other solutions.
+As a rule of thumb, including code snippets (except for defining an external API)
+is likely too low level. -->
+
+A WIP implementation can be found [here](https://github.com/ethereum-optimism/optimism/pull/12057).
+The specs can be found [here](https://github.com/ethereum-optimism/specs/tree/17ef36cdc3bb9893b206a93464122d56730d30fb/specs/protocol/holocene).
+
+Similar to the L1 MCP project, we move all of the network specific configuration out of the individual contracts
+themselves and instead place all of it in a single place. The contracts will make a `CALL` rather than using
+`sload` to read the values. These values will be sourced from L1 via deposit transactions that come from
+the `SystemConfig.initialize` call. We need to make sure that the max deposit gas limit is at least able to
+fullfill these deposit transactions.
+
+The general flow is as follows:
+
+```mermaid
+graph LR
+  subgraph L1
+  SystemConfig -- "setConfig(uint8,bytes)" --> OptimismPortal
+  end
+  subgraph L2
+  L1Block
+  BaseFeeVault -- "baseFeeVaultConfig()(address,uint256,uint8)" --> L1Block
+  SequencerFeeVault -- "sequencerFeeVaultConfig()(address,uint256,uint8)" --> L1Block
+  L1FeeVault -- "l1FeeVaultConfig()(address,uint256,uint8)" --> L1Block
+  L2CrossDomainMessenger -- "l1CrossDomainMessenger()(address)" --> L1Block
+  L2StandardBridge -- "l1StandardBridge()(address)" --> L1Block
+  L2ERC721Bridge -- "l1ERC721Bridge()(address)" --> L1Block
+  OptimismMintableERC721Factory -- "remoteChainId()(uint256)" --> L1Block
+  end
+  OptimismPortal -- "setConfig(uint8,bytes)" --> L1Block
+```
+
+This is taken from the [specs](https://github.com/ethereum-optimism/specs/blob/17ef36cdc3bb9893b206a93464122d56730d30fb/specs/protocol/holocene/predeploys.md) and misses the `L2ProxyAdmin`. The `L2ProxyAdmin` must also be deterministic
+and is explored in the following issue: https://github.com/ethereum-optimism/specs/issues/388. There is general
+consensus on using the `DEPOSITOR_ACCOUNT` as the owner.
+
+When we do a contract release, we commit to that bytecode as part of consensus. That is the bytecode used
+with deposit transactions doing upgrades to the network. In the L2 genesis creation script, we could have
+a library for each release of the predeploys. The genesis script would take the bytecode from the library if
+configured for a specific hardfork at genesis, otherwise it would use the compiled source code. This gives
+us a lot of flexibility and simplicity when it comes to being able to recreate an L2 genesis deterministically.
```

Comment:
Before committing to this, we should sync with slipper when he's back. In practice most chains will use op-deployer to generate their genesis, and I believe that would pull in the L2 contract bytecode of a release from a published docker image for that release

Reply:
This would only reduce the dependency of having the L1 deployments before and change nothing else about the op-deployer flow. What would you like to hear from Slipper specifically?

Reply:
> This would only reduce the dependency of having the L1 deployments before and change nothing else about the op-deployer flow.

Sorry, not following this sentence 😅

But basically I want to make sure we have a single canonical way of generating genesis. So assuming the genesis is done via op-deployer, you could either:
- Fetch bytecode from docker and use it (which I think is what the plan is), or
- Fetch bytecode from libraries in the repo that do not change

Reply:
Ah I see the difference we are talking about now. The solution that I propose around hardcoding the released bytecode into a library just makes generating historical genesis easy on any given commit. This prevents implicit deps, like foundry changing something and breaking the old script. Since we put the bytecode into Go code as part of consensus, it makes sense to put it in solidity too. The artifacts will contain the bytecode as well, the way the `L2Genesis` script is called should have minimal change if any. But I see your point with regards to looping in Slipper now, I don't think we need to commit to this when coming to consensus on the overall design.

Reply:
> hardcoding the released bytecode into a library

This will need to be documented in the release process

Reply:
Our release process is definitely getting complex. Related: https://www.notion.so/oplabs/Our-Release-Process-Is-Like-This-Title-A-Wee-Bit-Too-Long-and-Complex-a63f8a71f8454401b69b4877a9ad42df?pvs=4

We should consider a follow up design doc to automate as much as possible into a minimal set of `just commands`

Reply:
There is general consensus on the approach, see [here](https://discord.com/channels/1244729134312198194/1295923022389313646/1296537510012260374) for context

Reply:
Consensus: 
1. Bytecode goes into the library when a release is created.
  - needs automation to generate and CI to validate that the libraries bytecode always corresponds to the correct tag (possibly with a natspec tag). 
  - versioning.md will required an update, it will serve as a spec for the tooling.
2. The L2Genesis script can optionally pull from the artifacts

Reply:
- TODO: create a section outlining release considerations.
- TODO: get clarity about how to package up L2 contracts releases. How will alternative clients consume the L2 Genesis?

Reply:
Tracking this in the doc:
7cc8562dedeb